README.md
# ![TurkishCities](./public/assets/img/TurkishCities_yellow-on-darkblue_horizontal_914X343@3x-8.png)

<p align="center">
    <a href="https://rubygems.org/gems/turkish_cities">
        <img src="https://badge.fury.io/rb/turkish_cities.svg" alt="GemVersion"/></a>
    <a href="https://app.circleci.com/pipelines/github/sarslanoglu/turkish_cities">
        <img src="https://circleci.com/gh/sarslanoglu/turkish_cities.svg?style=shield" alt="BuildStatus"/></a>
    <a href="https://codecov.io/gh/sarslanoglu/turkish_cities">
        <img src="https://codecov.io/gh/sarslanoglu/turkish_cities/branch/master/graph/badge.svg?token=1SU7ZFI8ZJ"/></a>
    <a href="https://rubygems.org/gems/turkish_cities">
        <img src="https://img.shields.io/gem/dt/turkish_cities.svg" alt="DownloadCount"/></a>
    <a href="https://codeclimate.com/github/sarslanoglu/turkish_cities/maintainability">
        <img src="https://api.codeclimate.com/v1/badges/c1dbe0cef353b152956e/maintainability" alt="Maintainability"/></a>
    <a href="https://houndci.com">
        <img src="https://img.shields.io/badge/Protected_by-Hound-a873d1.svg" alt="Hound"/></a>
    <a href="https://opensource.org/licenses/MIT">
        <img src="https://img.shields.io/github/license/sarslanoglu/turkish_cities.svg" alt="License"/></a>
</p>

**Turkish Cities** is a Ruby gem which makes listing and finding Turkish cities easy. Search can be via name, post code, plate number, district name etc. Also calculate your travel distance and travel time between cities via distance methods.
https://rubygems.org/gems/turkish_cities


## Table of Contents
<details>
    <summary>Table of Contents</summary>
    <ul>
        <li><a href="#installation">Installation</a></li>
        <li><a href="#localization">Localization</a></li>
        <li>
            <a href="#documentation">Documentation</a>
            <ul>
                <li><a href="#finding-city-name-by-plate-number">Finding city name by plate number</a></li>
                <li><a href="#finding-city-name-by-phone-code">Finding city name by phone code</a></li>
                <li><a href="#finding-plate-number-by-city-name">Finding plate number by city name</a></li>
                <li><a href="#finding-phone-code-by-city-name">Finding phone code by city name</a></li>
                <li>
                    <a href="#listing-all-cities">Listing all cities</a>
                    <ul>
                        <li><a href="#listing-all-cities-with-only-name">Listing all cities with only name</a></li>
                        <li><a href="#listing-all-cities-with-other-parameters">Listing all cities with other parameters</a></li>
                    </ul>
                </li>
                <li><a href="#listing-all-districts-of-given-city">Listing all districts of given city</a></li>
                <li><a href="#listing-all-subdistricts-of-given-city-and-district">Listing all subdistricts of given city and district</a></li>
                <li>
                    <a href="#listing-all-neighborhoods-of-given-city-and-district">Listing all neighborhoods of given city and district</a>
                    <ul>
                        <li><a href="#with-subdistrict-info">With subdistrict info</a></li>
                        <li><a href="#without-subdistrict-info">Without subdistrict info</a></li>
                    </ul>
                </li>
                <li><a href="#finding-city-district-and-subdistrict-name-by-postcode">Finding city, district and subdistrict name by postcode</a></li>
                <li>
                    <a href="#finding-city-name-with-elevation-data">Finding city name with elevation data</a>
                    <ul>
                        <li><a href="#below-given">Below given</a></li>
                        <li><a href="#above-given">Above given</a></li>
                        <li><a href="#between-search">Between search</a></li>
                    </ul>
                </li>
                <li>
                    <a href="#finding-city-name-with-population-data">Finding city name with population data</a>
                    <ul>
                        <li><a href="#exact-search">Exact search</a></li>
                        <li><a href="#below-given">Below given</a></li>
                        <li><a href="#above-given">Above given</a></li>
                        <li><a href="#between-search">Between search</a></li>
                    </ul>
                </li>
                <li>
                    <a href="#finding-travel-distance-and-time-estimation-between-two-cities">Finding travel distance and time estimation between two cities</a>
                    <ul>
                        <li><a href="#by-land">By land</a></li>
                        <li><a href="#by-air">By air</a></li>
                    </ul>
                </li>
            </ul>
        </li>
        <li><a href="#data-sources">Data sources</a></li>
        <li><a href="#compatibility">Compatibility</a></li>
        <li><a href="#contributing">Contributing</a></li>
        <li><a href="#changelog">Changelog</a></li>
        <li><a href="#copyright">Copyright</a></li>
    </ul>
</details>

## Installation

Installation is pretty standard:

```sh
$ gem install turkish_cities
```

or with using Bundler, add this line to your Gemfile

```rb
gem 'turkish_cities'
```

and run

```sh
$ bundle
```

## Localization

Currently gem supports two different languages, Turkish and English. English is default, if no action taken.

Language can be easily changed by ```change_locale``` method. Valid parameters are 'tr' for Turkish and 'en' for English.

```rb
TurkishCities.change_locale('tr') # => "Dil Türkçe olarak ayarlandı."
```

After changing language all errors and result sets will be in desired language

```rb
TurkishCities.list_districts('Eskişehirrr') # => "'Eskişehirrr' ile bir şehir bulunamadı"
TurkishCities.distance_between('kirsehir', 'Ordu', 'land')
# => [533, "Kırşehir ile Ordu arasındaki karayolu mesafesi 533 km"]
```

## Documentation

With using irb just require gem and start using

```rb
require 'turkish_cities'
```

### Finding city name by plate number

There are 81 cities in Turkey. By calling a plate number between 1-81 will give city_name.

<details>
    <summary>Expand</summary>

```rb
TurkishCities.find_name_by_plate_number(26)   # => "Eskişehir"
TurkishCities.find_name_by_plate_number(7)    # => "Antalya"
TurkishCities.find_name_by_plate_number(0007) # => "Antalya"
TurkishCities.find_name_by_plate_number(43.0) # => "Kütahya"
TurkishCities.find_name_by_plate_number('78') # => "Karabük"
TurkishCities.find_name_by_plate_number(100)  # => 'Given value [100] is outside bounds of 1 to 81.'
```
</details>

### Finding city name by phone code

There are 82 phone codes for each city in Turkey (Istanbul is an exception with different phone codes for Anatolian side and European side). By calling a phone code between 212-488 will give city_name. All phone codes ending with even numbers. So sending an odd number will give an error without searching.

<details>
    <summary>Expand</summary>

```rb
TurkishCities.find_name_by_phone_code(312)    # => "Ankara"
TurkishCities.find_name_by_phone_code(242)    # => "Antalya"
TurkishCities.find_name_by_phone_code(000222) # => "Eskişehir"
TurkishCities.find_name_by_phone_code(274.0)  # => "Kütahya"
TurkishCities.find_name_by_phone_code('212')  # => "İstanbul"
TurkishCities.find_name_by_phone_code(216)    # => "İstanbul"
TurkishCities.find_name_by_phone_code(360)    # => 'Couldn't find city name with phone code 360'
TurkishCities.find_name_by_phone_code(0)      # => 'Given value [0] is outside bounds of 212 to 488.'
TurkishCities.find_name_by_phone_code(213)    # => 'Given value [213] must be an even number.'
```
</details>

### Finding plate number by city name

City name can be given case and turkish character insensitive.

<details>
    <summary>Expand</summary>

```rb
TurkishCities.find_plate_number_by_name('Ankara')     # => 6
TurkishCities.find_plate_number_by_name('Eskişehir')  # => 26
TurkishCities.find_plate_number_by_name('Canakkale')  # => 17
TurkishCities.find_plate_number_by_name('Istanbul')   # => 34
TurkishCities.find_plate_number_by_name('kirsehir')   # => 40
TurkishCities.find_plate_number_by_name('falansehir') # => "Couldn't find city name with 'falansehir'"
```
</details>

### Finding phone code by city name

City name can be given case and turkish character insensitive.

<details>
    <summary>Expand</summary>

```rb
TurkishCities.find_phone_code_by_name('Ankara')     # => 312
TurkishCities.find_phone_code_by_name('Eskişehir')  # => 222
TurkishCities.find_phone_code_by_name('Canakkale')  # => 286
TurkishCities.find_phone_code_by_name('Istanbul')   # => [212, 216]
TurkishCities.find_phone_code_by_name('kirsehir')   # => 386
TurkishCities.find_phone_code_by_name('filansehir') # => "Couldn't find city name with 'filansehir'"
```
</details>

### Listing all cities

By default cities will be listed by their plate number ascending.

<details>
    <summary>Expand</summary>

```rb
TurkishCities.list_cities # => ["Adana", "Adıyaman" ... "Kilis", "Osmaniye", "Düzce"]
```
</details>

### Listing all cities with only name

While listing cities three additional parameters can be sent ```alphabetically_sorted```, ```metropolitan_municipality``` and ```region```. All parameters can be sent separately and together.

<details>
    <summary>Expand</summary>

```rb
TurkishCities.list_cities({ alphabetically_sorted: true })
# => ["Adana", "Adıyaman" ... "Yalova", "Yozgat", "Zonguldak"]
TurkishCities.list_cities({ metropolitan_municipality: true })
# => ["Adana", "Ankara" ... "Trabzon", "Şanlıurfa", "Van"]
TurkishCities.list_cities({ region: 'Karadeniz' })
# => ["Amasya", "Artvin" ... "Bartın", "Karabük", "Düzce"]
TurkishCities.list_cities({ alphabetically_sorted: true, region: 'Karadeniz' })
# => ["Amasya", "Artvin" ... "Tokat", "Trabzon", "Zonguldak"]
TurkishCities.list_cities({ metropolitan_municipality: true, region: 'Karadeniz' })
# => ["Ordu", "Samsun", "Trabzon"]
TurkishCities.list_cities({ alphabetically_sorted: true, metropolitan_municipality: true })
# => ["Adana", "Ankara" ... "Tekirdağ", "Trabzon", "Van"]
```
</details>

### Listing all cities with other parameters

While listing cities ```with``` parameter can be used for listing cities with other attributes. These parameters are ```alphabetically_sorted```, ```metropolitan_municipality``` and ```region```. All parameters can be sent separately and together.

<details>
    <summary>Expand</summary>

```rb
TurkishCities.list_cities({ alphabetically_sorted: true, with: { phone_code: true } })
# => [{:name=>"Adana", :phone_code=>322}, {:name=>"Adıyaman", :phone_code=>416}, {:name=>"Afyon", :phone_code=>272} .
# .. {:name=>"Yozgat", :phone_code=>354}, {:name=>"Zonguldak", :phone_code=>372}]
TurkishCities.list_cities({ with: { plate_number: true } })
# => [{:name=>"Adana", :plate_number=>1}, {:name=>"Adıyaman", :plate_number=>2}, {:name=>"Afyon", :plate_number=>3} .
# .. {:name=>"Osmaniye", :plate_number=>80}, {:name=>"Düzce", :plate_number=>81}]
TurkishCities.list_cities({ metropolitan_municipality: true, region: 'Karadeniz', with: { all: true } })
# => [{:name=>"Ordu", :plate_number=>52, :phone_code=>452, :metropolitan_municipality_since=>2012,
#      :region=>"Karadeniz"},{:name=>"Samsun", :plate_number=>55, :phone_code=>362,
#      :metropolitan_municipality_since=>1993, :region=>"Karadeniz"}, {:name=>"Trabzon", :plate_number=>61,
#      :phone_code=>462, :metropolitan_municipality_since=>2012, :region=>"Karadeniz"}]
```
</details>

### Listing all districts of given city

City name can be given case and turkish character insensitive. Listing of districts are alphabetically sorted.

<details>
    <summary>Expand</summary>

```rb
TurkishCities.list_districts('Ankara')
# => ["Akyurt", "Altındağ" ... "Sincan", "Şereflikoçhisar", "Yenimahalle"]
TurkishCities.list_districts('Eskişehir')
# => ["Alpu", "Beylikova", "Çifteler" ... "Sivrihisar", "Tepebaşı"]
TurkishCities.list_districts('Canakkale')
# => ["Ayvacık", "Bayramiç" ... "Lapseki", "Merkez", "Yenice"]
TurkishCities.list_districts('Bayburt')
# => ["Aydıntepe", "Demirözü", "Merkez"]
TurkishCities.list_districts('Istanbul').length
# => 39
TurkishCities.list_districts('filansehir')
# => "Couldn't find city name with 'filansehir'"
```
</details>

### Listing all subdistricts of given city and district

City name can be given case and turkish character insensitive. District name should be case and turkish character sensitive.(Correct district names can be obtained by ```list_districts``` method.) Listing of subdistricts are alphabetically sorted.

<details>
    <summary>Expand</summary>

```rb
TurkishCities.list_subdistricts('Adana', 'Seyhan')
# => ["Akkapı", "Denizli" ... "Yeşiloba", "Yeşilyurt", "Ziyapaşa"]
TurkishCities.list_subdistricts('Eskişehir', 'Odunpazarı')
# => ["Alanönü", "Arifiye", "Büyükdere" ... "Vişnelik", "Yenikent"]
TurkishCities.list_subdistricts('Istanbul', 'Beşiktaş')
# => ["Abbasağa", "Akatlar", "Arnavutköy", "Bebek", "Etiler", "Gayrettepe", "Levazım", "Levent", "Ortaköy", "Türkali"]
TurkishCities.list_subdistricts('filansehir', 'Beşiktaş')
# => "Couldn't find district name with 'Beşiktaş' of 'filansehir'"
TurkishCities.list_subdistricts('İstanbul', 'Kadılarköyü')
# => "Couldn't find district name with 'Kadılarköyü' of 'İstanbul'"
```

</details>

### Listing all neighborhoods of given city and district

### With subdistrict info

City name can be given case and turkish character insensitive. District name and subdistrict name should be case and turkish character sensitive.(Correct district names can be obtained by the ```list_districts``` method. Correct subdistrict names can be obtained by ```list_subdistricts``` method.) Listings of neighborhoods are alphabetically sorted.

<details>
    <summary>Expand</summary>

```rb
TurkishCities.list_neighborhoods('Adana', 'Seyhan', 'Emek')
# => ["Emek Mah", "Ova Mah", "Şakirpaşa Mah", "Uçak Mah"]
TurkishCities.list_neighborhoods('Eskişehir', 'Odunpazarı', 'Büyükdere')
# => ["Büyükdere Mah", "Göztepe Mah", "Gültepe Mah", "Yıldıztepe Mah"]
TurkishCities.list_neighborhoods('Istanbul', 'Beşiktaş', 'Gayrettepe')
# => ["Balmumcu Mah", "Dikilitaş Mah", "Gayrettepe Mah", "Yıldız Mah"]
TurkishCities.list_neighborhoods('Adana', 'Beşiktaş', 'Emek')
# => "Couldn't find district name with 'Beşiktaş' of 'Adana'"
TurkishCities.list_neighborhoods('Eskişehir', 'Odunpazarı', 'Büyükkkkkdere')
# => "Couldn't find subdistrict with 'Büyükkkkkdere' of 'Odunpazarı'/'Eskişehir'"
```
</details>

### Without subdistrict info

Also ```list_neighborhoods``` can work without subdistrict information. This time neighborhoods result will be larger based on searched city and district.

City name can be given case and turkish character insensitive. District name should be case and turkish character sensitive.(Correct district names can be obtained by the ```list_districts``` method.) Listings of neighborhoods are alphabetically sorted.

<details>
    <summary>Expand</summary>

```rb
TurkishCities.list_neighborhoods('Adana', 'Seyhan')
# => ["2000 Evler Mah", "Ahmet Remzi Yüreğir Mah", "Akkapı Mah" ... "Zeytinli Mah", "Ziyapaşa Mah"]
TurkishCities.list_neighborhoods('Eskişehir', 'Odunpazarı')
# => ["71 Evler Mah", "75. Yıl Osb Mah" ... "Yukarıkalabak Mah", "Yürükkaracaören Mah", "Yürükkırka Mah"]
TurkishCities.list_neighborhoods('Istanbul', 'Beşiktaş')
# => ["Abbasağa Mah", "Akat Mah", "Arnavutköy Mah" ... "Türkali Mah", "Ulus Mah", "Vişnezade Mah", "Yıldız Mah"]
TurkishCities.list_neighborhoods('İstanbul', 'filanmevki')
# => "Couldn't find district name with 'filanmevki' of 'İstanbul'"
```
</details>

### Finding city, district and subdistrict name by postcode

In Turkey postcodes are uniq for subdistricts. Basically when a postcode is search through ```find_by_postcode``` it will give city, district and subdistrict information of postcode if valid and found

<details>
    <summary>Expand</summary>

```rb
TurkishCities.find_by_postcode(34380)    # => ["İstanbul", "Şişli", "Cumhuriyet"]
TurkishCities.find_by_postcode(34433)    # => ["İstanbul", "Beyoğlu", "Cihangir"]
TurkishCities.find_by_postcode('26040')  # => ["Eskişehir", "Odunpazarı", "Büyükdere"]
TurkishCities.find_by_postcode(34382)    # => "Couldn't find any subdistrict with postcode '34382'"
TurkishCities.find_by_postcode('100000') # => Given value [100000] is outside bounds of 1010 to 81952.
```
</details>

### Finding city name with elevation data

The lowest possible altitude for a city in Turkey is 4 meters with Kocaeli. The highest possible altitude is 1923 meters with Erzurum. Being far outside of the range may give errors.

### Below given

Search city name with below elevation

<details>
    <summary>Expand</summary>

```rb
TurkishCities.find_by_elevation('below', 10)
# => ["İzmir", "Kocaeli", "Yalova"]
TurkishCities.find_by_elevation('below', 50)
# => ["Adana", "Çanakkale" ... "Bartın", "Yalova"]
TurkishCities.find_by_elevation('below', 500)
# => ["Adana", "Amasya" ... "Karabük", "Osmaniye", "Düzce"]
TurkishCities.find_by_elevation('below', 2213)
# => ["Adana", "Adıyaman" ... "Kilis", "Osmaniye", "Düzce"]
TurkishCities.find_by_elevation('woleb', 213)
# => ArgumentError: "Elevation type 'woleb' is unsupported"
TurkishCities.find_by_elevation('below', 1)
# => RangeError: Given value [1] is outside bounds of 3 to Infinity
```
</details>

### Above given

Search city name with above elevation

<details>
    <summary>Expand</summary>

```rb
TurkishCities.find_by_elevation('above', 0)
# => ["Adana", "Adıyaman" ... "Osmaniye", "Düzce"]
TurkishCities.find_by_elevation('above', 100)
# => ["Adıyaman", "Afyon" ... "Osmaniye", "Düzce"]
TurkishCities.find_by_elevation('above', 500)
# => ["Adıyaman", "Afyon" ... "Iğdır", "Kilis"]
TurkishCities.find_by_elevation('above', 1750)
# => ["Erzurum", "Hakkari", "Kars", "Ardahan"]
TurkishCities.find_by_elevation('abov', 1000)
# => ArgumentError: "Elevation type 'abov' is unsupported"
TurkishCities.find_by_elevation('above', 2000)
# => RangeError: Given value [2000] is outside bounds of 0 to 1924
```
</details>

### Between search

Search city name with between elevation

<details>
    <summary>Expand</summary>

```rb
TurkishCities.find_by_elevation('between', 50, 100)
# => ["Antalya", "Aydın", "Hatay", "Manisa"]
TurkishCities.find_by_elevation('between', 100, 250)
# => ["Balıkesir", "Bursa", "Kırklareli", "Osmaniye", "Düzce"]
TurkishCities.find_by_elevation('between', 250, 1000)
# => ["Adıyaman", "Amasya" ... "Karabük", "Kilis"]
TurkishCities.find_by_elevation('between', 0, 2000)
# => ["Adana", "Adıyaman" ... "Osmaniye", "Düzce"]
TurkishCities.find_by_elevation('between', 2000, 5000)
# => []
```
</details>

### Finding city name with population data

The least populated city in Turkey is Tunceli with 83645. The highest populated is Istanbul with 15840900. Being far outside of the range may give errors.

### Exact search

Search city name with exact population

<details>
    <summary>Expand</summary>

```rb
TurkishCities.find_by_population('exact', 15840900)
# => ["İstanbul"]
TurkishCities.find_by_population('exact', 104320)
# => "Couldn't find any city with population data"
TurkishCities.find_by_population('exatc', 2130432)
# => "Population type 'exatc' is unsupported"
TurkishCities.find_by_population('exact', 10432)
# => Given value [10432] is outside bounds of 83644 to 15840901
TurkishCities.find_by_population('exact', 22130432)
# => Given value [22130432] is outside bounds of 83644 to 15840901
```
</details>

### Below given

Search city name with below population

<details>
    <summary>Expand</summary>

```rb
TurkishCities.find_by_population('below', 86000)
# => ["Tunceli", "Bayburt"]
TurkishCities.find_by_population('below', 500000)
# => ["Amasya", "Artvin" ... "Kilis", "Düzce"]
TurkishCities.find_by_population('below', 1000000)
# => ["Adıyaman", "Afyon" ... "Kilis", "Osmaniye", "Düzce"]
TurkishCities.find_by_population('below', 22130432)
# => ["Adana", "Adıyaman" ... "Kilis", "Osmaniye", "Düzce"]
TurkishCities.find_by_population('woleb', 2130432)
# => "Population type 'woleb' is unsupported"
TurkishCities.find_by_population('below', 10432)
# => Given value [10432] is outside bounds of 83644 to Infinity
```
</details>

### Above given

Search city name with above population

<details>
    <summary>Expand</summary>

```rb
TurkishCities.find_by_population('above', 860000)
# => ["Adana", "Ankara" ... "Şanlıurfa", "Van"]
TurkishCities.find_by_population('above', 2500000)
# => ["Ankara", "Antalya", "Bursa", "İstanbul", "İzmir"]
TurkishCities.find_by_population('above', 5000000)
# => ["Ankara", "İstanbul"]

CHANGELOG.md
# Changelog

## Unreleased version

### New Features

* [#127](https://github.com/sarslanoglu/turkish_cities/issues/127): Add control for missing chars on ```sort_alphabetically``` method

### Bug fixes

* [#129](https://github.com/sarslanoglu/turkish_cities/issues/129): NoMethodError undefined method `values'. Update Errno::ENOENT message

* [#124](https://github.com/sarslanoglu/turkish_cities/issues/124): ArgumentError: comparison of Array with Array failed. Add support for "()" characters

* [#122](https://github.com/sarslanoglu/turkish_cities/issues/122): ArgumentError: comparison of Array with Array failed. Add support for "." character

## 0.7.0 (2022-08-03)

### New features

* [#14](https://github.com/sarslanoglu/turkish_cities/issues/14): Add city altitude data and ```find_by_elevation``` method

### Bug fixes

* [#116](https://github.com/sarslanoglu/turkish_cities/issues/116): translation missing error on Rails apps

## 0.6.1 (2022-07-18)

### Bug fixes

* [#113](https://github.com/sarslanoglu/turkish_cities/issues/113): Fix no implicit conversion of Array into String error

## 0.6.0 (2022-02-26)

### New features

* [#12](https://github.com/sarslanoglu/turkish_cities/issues/12): Add city population data and ```find_by_population``` method

### Changes

* [#103](https://github.com/sarslanoglu/turkish_cities/issues/103): Update city/district/subdistrict informations to 2022

## 0.5.0 (2021-06-06)

### New features

* [#68](https://github.com/sarslanoglu/turkish_cities/issues/68): Add localization for all method responses and errors

## 0.4.0 (2021-04-25)

### New features

* [#66](https://github.com/sarslanoglu/turkish_cities/issues/66): Add ```distance_between_land``` method for calculating land travel distances between cities

## 0.3.0 (2020-11-22)

### New features

* [#13](https://github.com/sarslanoglu/turkish_cities/issues/13): Add ```distance``` class and ```distance_between``` method for calculating air travel distance and time estimation between cities

### Changes

* [#50](https://github.com/sarslanoglu/turkish_cities/issues/50): Additional improvements on `District` class methods

* [#45](https://github.com/sarslanoglu/turkish_cities/issues/45): Refactor `District` class

## 0.2.1 (2020-06-29)

### Bug fixes

* [#38](https://github.com/sarslanoglu/turkish_cities/issues/38): ```NoMethodError``` on ```list_neighborhoods``` when called with wrong subdistrict name

### Changes

* [#39](https://github.com/sarslanoglu/turkish_cities/issues/39): May maintenance of gem - 2020-05-24

## 0.2.0 (2020-05-18)

### New features

* [#36](https://github.com/sarslanoglu/turkish_cities/issues/36): Add ```postcode``` class and ```find_by_postcode``` method for finding city, district and subdistrict information by given postcode

* [#33](https://github.com/sarslanoglu/turkish_cities/issues/33): Add ```list_neighborhoods``` method for listing neighborhoods and fix some wrong district names

* [#29](https://github.com/sarslanoglu/turkish_cities/issues/29): Add ```district``` class and ```list_subdistricts``` method for listing subdistricts

* [#27](https://github.com/sarslanoglu/turkish_cities/issues/27): Add all subdistricts and neighborhoods of all Turkish cities

* [#25](https://github.com/sarslanoglu/turkish_cities/issues/25): Add district data to each city at yaml file and add new method ```list_districts``` for listing disctricts of cities alphabetically sorted

### Changes

* [#24](https://github.com/sarslanoglu/turkish_cities/issues/24): Add listing support with ```metropolitan_municipality_since, phone_code, plate_number, region``` to ```listing_cities``` method

## 0.1.3 (2020-04-20)

### New features

* [#22](https://github.com/sarslanoglu/turkish_cities/issues/22): Add region data to cities.yaml and update ```list_cities``` method to support regions

* [#20](https://github.com/sarslanoglu/turkish_cities/issues/20): Add city finding with phone code and vice versa. ```find_name_by_phone_code``` and ```find_phone_code_by_name``` methods are added

### Bug fixes

* [#18](https://github.com/sarslanoglu/turkish_cities/issues/18): Fix yaml file read error while deploying apps to Heroku

## 0.1.2 (2020-04-13)

### New features

* [#3](https://github.com/sarslanoglu/turkish_cities/issues/3): Change city_list data to yaml file format

### Bug fixes

* [#2](https://github.com/sarslanoglu/turkish_cities/issues/2): Fix capital Turkish characters bug on cities with capital 'I' letter

### Changes

* [#15](https://github.com/sarslanoglu/turkish_cities/issues/15): Handle error messages at ```find_name_by_plate_number``` and ```find_plate_number_by_name``` methods

## 0.1.1 (2020-03-31)

### Changes

* Fix homepage problem on .gemspec file


lib/turkish_cities/data/city_list.rb
# frozen_string_literal: true

CITY_LIST = {
  1 => { name: 'Adana',
         phone_code: 322,
         metropolitan_municipality_since: 1986 },
  2 => { name: 'Adıyaman',
         phone_code: 416,
         metropolitan_municipality_since: nil },
  3 => { name: 'Afyon',
         phone_code: 272,
         metropolitan_municipality_since: nil },
  4 => { name: 'Ağrı',
         phone_code: 472,
         metropolitan_municipality_since: nil },
  5 => { name: 'Amasya',
         phone_code: 358,
         metropolitan_municipality_since: nil },
  6 => { name: 'Ankara',
         phone_code: 312,
         metropolitan_municipality_since: 1984 },
  7 => { name: 'Antalya',
         phone_code: 242,
         metropolitan_municipality_since: 1993 },
  8 => { name: 'Artvin',
         phone_code: 466,
         metropolitan_municipality_since: nil },
  9 => { name: 'Aydın',
         phone_code: 256,
         metropolitan_municipality_since: 2012 },
  10 => { name: 'Balıkesir',
          phone_code: 266,
          metropolitan_municipality_since: 2012 },
  11 => { name: 'Bilecik',
          phone_code: 228,
          metropolitan_municipality_since: nil },
  12 => { name: 'Bingöl',
          phone_code: 426,
          metropolitan_municipality_since: nil },
  13 => { name: 'Bitlis',
          phone_code: 434,
          metropolitan_municipality_since: nil },
  14 => { name: 'Bolu',
          phone_code: 374,
          metropolitan_municipality_since: nil },
  15 => { name: 'Burdur',
          phone_code: 248,
          metropolitan_municipality_since: nil },
  16 => { name: 'Bursa',
          phone_code: 224,
          metropolitan_municipality_since: 1987 },
  17 => { name: 'Çanakkale',
          phone_code: 286,
          metropolitan_municipality_since: nil },
  18 => { name: 'Çankırı',
          phone_code: 376,
          metropolitan_municipality_since: nil },
  19 => { name: 'Çorum',
          phone_code: 364,
          metropolitan_municipality_since: nil },
  20 => { name: 'Denizli',
          phone_code: 258,
          metropolitan_municipality_since: 2012 },
  21 => { name: 'Diyarbakır',
          phone_code: 412,
          metropolitan_municipality_since: 1993 },
  22 => { name: 'Edirne',
          phone_code: 284,
          metropolitan_municipality_since: nil },
  23 => { name: 'Elazığ',
          phone_code: 424,
          metropolitan_municipality_since: nil },
  24 => { name: 'Erzincan',
          phone_code: 446,
          metropolitan_municipality_since: nil },
  25 => { name: 'Erzurum',
          phone_code: 442,
          metropolitan_municipality_since: 1993 },
  26 => { name: 'Eskişehir',
          phone_code: 222,
          metropolitan_municipality_since: 1993 },
  27 => { name: 'Gaziantep',
          phone_code: 342,
          metropolitan_municipality_since: 1987 },
  28 => { name: 'Giresun',
          phone_code: 454,
          metropolitan_municipality_since: nil },
  29 => { name: 'Gümüşhane',
          phone_code: 456,
          metropolitan_municipality_since: nil },
  30 => { name: 'Hakkari',
          phone_code: 438,
          metropolitan_municipality_since: nil },
  31 => { name: 'Hatay',
          phone_code: 326,
          metropolitan_municipality_since: 2012 },
  32 => { name: 'Isparta',
          phone_code: 246,
          metropolitan_municipality_since: nil },
  33 => { name: 'Mersin',
          phone_code: 324,
          metropolitan_municipality_since: 1993 },
  34 => { name: 'İstanbul',
          phone_code: [212, 216],
          metropolitan_municipality_since: 1984 },
  35 => { name: 'İzmir',
          phone_code: 232,
          metropolitan_municipality_since: 1984 },
  36 => { name: 'Kars',
          phone_code: 474,
          metropolitan_municipality_since: nil },
  37 => { name: 'Kastamonu',
          phone_code: 366,
          metropolitan_municipality_since: nil },
  38 => { name: 'Kayseri',
          phone_code: 352,
          metropolitan_municipality_since: 1988 },
  39 => { name: 'Kırklareli',
          phone_code: 288,
          metropolitan_municipality_since: nil },
  40 => { name: 'Kırşehir',
          phone_code: 386,
          metropolitan_municipality_since: nil },
  41 => { name: 'Kocaeli',
          phone_code: 262,
          metropolitan_municipality_since: 1993 },
  42 => { name: 'Konya',
          phone_code: 332,
          metropolitan_municipality_since: 1987 },
  43 => { name: 'Kütahya',
          phone_code: 274,
          metropolitan_municipality_since: nil },
  44 => { name: 'Malatya',
          phone_code: 422,
          metropolitan_municipality_since: 2012 },
  45 => { name: 'Manisa',
          phone_code: 236,
          metropolitan_municipality_since: 2012 },
  46 => { name: 'Kahramanmaraş',
          phone_code: 344,
          metropolitan_municipality_since: 2012 },
  47 => { name: 'Mardin',
          phone_code: 482,
          metropolitan_municipality_since: 2012 },
  48 => { name: 'Muğla',
          phone_code: 252,
          metropolitan_municipality_since: 2012 },
  49 => { name: 'Muş',
          phone_code: 436,
          metropolitan_municipality_since: nil },
  50 => { name: 'Nevşehir',
          phone_code: 384,
          metropolitan_municipality_since: nil },
  51 => { name: 'Niğde',
          phone_code: 388,
          metropolitan_municipality_since: nil },
  52 => { name: 'Ordu',
          phone_code: 452,
          metropolitan_municipality_since: 2012 },
  53 => { name: 'Rize',
          phone_code: 464,
          metropolitan_municipality_since: nil },
  54 => { name: 'Sakarya',
          phone_code: 264,
          metropolitan_municipality_since: 2000 },
  55 => { name: 'Samsun',
          phone_code: 362,
          metropolitan_municipality_since: 1993 },
  56 => { name: 'Siirt',
          phone_code: 484,
          metropolitan_municipality_since: nil },
  57 => { name: 'Sinop',
          phone_code: 368,
          metropolitan_municipality_since: nil },
  58 => { name: 'Sivas',
          phone_code: 346,
          metropolitan_municipality_since: nil },
  59 => { name: 'Tekirdağ',
          phone_code: 282,
          metropolitan_municipality_since: 2012 },
  60 => { name: 'Tokat',
          phone_code: 356,
          metropolitan_municipality_since: nil },
  61 => { name: 'Trabzon',
          phone_code: 462,
          metropolitan_municipality_since: 2012 },
  62 => { name: 'Tunceli',
          phone_code: 428,
          metropolitan_municipality_since: nil },
  63 => { name: 'Şanlıurfa',
          phone_code: 414,
          metropolitan_municipality_since: 2012 },
  64 => { name: 'Uşak',
          phone_code: 276,
          metropolitan_municipality_since: nil },
  65 => { name: 'Van',
          phone_code: 432,
          metropolitan_municipality_since: 2012 },
  66 => { name: 'Yozgat',
          phone_code: 354,
          metropolitan_municipality_since: nil },
  67 => { name: 'Zonguldak',
          phone_code: 372,
          metropolitan_municipality_since: nil },
  68 => { name: 'Aksaray',
          phone_code: 382,
          metropolitan_municipality_since: nil },
  69 => { name: 'Bayburt',
          phone_code: 458,
          metropolitan_municipality_since: nil },
  70 => { name: 'Karaman',
          phone_code: 338,
          metropolitan_municipality_since: nil },
  71 => { name: 'Kırıkkale',
          phone_code: 318,
          metropolitan_municipality_since: nil },
  72 => { name: 'Batman',
          phone_code: 488,
          metropolitan_municipality_since: nil },
  73 => { name: 'Şırnak',
          phone_code: 486,
          metropolitan_municipality_since: nil },
  74 => { name: 'Bartın',
          phone_code: 378,
          metropolitan_municipality_since: nil },
  75 => { name: 'Ardahan',
          phone_code: 478,
          metropolitan_municipality_since: nil },
  76 => { name: 'Iğdır',
          phone_code: 476,
          metropolitan_municipality_since: nil },
  77 => { name: 'Yalova',
          phone_code: 226,
          metropolitan_municipality_since: nil },
  78 => { name: 'Karabük',
          phone_code: 370,
          metropolitan_municipality_since: nil },
  79 => { name: 'Kilis',
          phone_code: 348,
          metropolitan_municipality_since: nil },
  80 => { name: 'Osmaniye',
          phone_code: 328,
          metropolitan_municipality_since: nil },
  81 => { name: 'Düzce',
          phone_code: 380,
          metropolitan_municipality_since: nil }
}.freeze
